### PR TITLE
Pubsub 2

### PR DESCRIPTION
**Pull request by connor4312**
* **Breaking Change**: New() now takes a ConnectionParam value rather than a pointer.
 * **Breaking Change**: GetState() now returns a uint8 rather than a user-defined type, for greater compatibility with [fsm](https://github.com/WatchBeam/fsm). <small>_mumble mumble generics_</small>
 * Fix potential data races on the internal subscription registry.
 * Fix potential data race resulting in subscription duplication during multiple reconnections.
 * Allow specification of connection timeout (deadlines).
 * Allow specification of reconnection policies.
 * Cause subscription, unsubscriptions, and teardowns to happen more quickly.
 * Improve events system for increased flexibility.
 * Significantly improve conciseness and speed.

### CODE REVIEW

File: pubsub/client.go [hunk 1]
```diff
@@ -161,116 +132,131 @@ func (c *Client) Listener(kind ListenerType, event string) *Listener {
...
-		if ev := <-c.Events; ev.Type == event {
-			return ev
-		}
+// Tries to reconnect to pubsub, looping until we're able to do so
+// successfully. This must be called to activate the client.
+func (c *Client) Connect() {
+	for c.GetState() == DisconnectedState {
+		go c.resubscribe()
+		c.doConnection()
+		time.Sleep(c.policy.Next())
```
Comment (ttaylorr) [inline pubsub/client.go]: Perhaps a `time.Ticker` would work well here?

**Comment by ttaylorr** [comment]
One minor comment, otherwise great refactoring :smile:

File: pubsub/client.go (hunk 1 above)
Comment (connor4312) [inline pubsub/client.go]: I'm not sure that's really appropriate here. It's a single delay, not an interval.

File: pubsub/client.go [hunk 2]
```diff
@@ -113,35 +75,44 @@ type action uint8
...
-		address:     conn.Address,
-		password:    conn.Password,
-		subscribed:  map[string][]*Listener{},
-		actionQueue: make(chan task, 10),
-		stateLock:   new(sync.Mutex),
-		pubsubLock:  new(sync.Mutex),
+		eventEmitter: newEventEmitter(),
+		state:        blueprint.Machine(),
+		stateLock:    new(sync.Mutex),
+		conn:         conn,
```
Comment (janeczku) [inline pubsub/client.go]: so much better :-)

File: readme.md [hunk 3]
```diff
@@ -20,7 +20,7 @@ import (
 func main() {
     // Create a new pubsub client. This will create and manage connections,
     // even if you disconnect.
-    c := pubsub.New(&pubsub.ConnectionParam{
+    c := pubsub.New(pubsub.ConnectionParam{
         Address: "127.0.0.1:6379",
         // optional password
         Password: "secret"
```
Comment (janeczku) [inline readme.md]: missing trailing comma
```go
Password: "secret",
```

**Comment by janeczku** [comment]
Great work :thumbsup:

**Pull request merged**